File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        auto dfs = [&](auto&& self,TreeNode* node)->TreeNode*{
            if(!node || node==p || node==q) return node;
            TreeNode* node1 = self(self,node->left);
            TreeNode* node2 = self(self,node->right);
            if(node1 && node2) return node;
            return node1?node1:node2;
        };

        return dfs(dfs,root);
    }
};

File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def f(node):
            if node is None or node==p or node==q:
                return node
            
            node1=f(node.left)
            node2=f(node.right)

            if node1 and node2:
                return node
                
            return node1 if node1 else node2

        return f(root)